File: backend/src/main/java/com/memeit/backend/dataclasses/Reaction.java
package com.memeit.backend.dataclasses;

import android.os.Parcel;
import android.os.Parcelable;

import com.google.gson.annotations.SerializedName;

/**
 * Created by Jv on 6/29/2018.
 */

public class Reaction implements Parcelable {
    protected Reaction(Parcel in) {
        type = in.readInt();
        memeID = in.readString();
        reactorID = in.readString();
    }

    public static final Creator<Reaction> CREATOR = new Creator<Reaction>() {
        @Override
        public Reaction createFromParcel(Parcel in) {
            return new Reaction(in);
        }

        @Override
        public Reaction[] newArray(int size) {
            return new Reaction[size];
        }
    };

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(type);
        dest.writeString(memeID);
        dest.writeString(reactorID);
    }

    public enum ReactionType{
        FUNNY(2f),VERY_FUNNY(3f),STUPID(-1f),ANGERING(-2F);

        private float score;

        ReactionType(float score) {
            this.score = score;
        }

        public float getScore() {
            return score;
        }
    }
    @SerializedName("type")
    private int type;
    @SerializedName("mid")
    private String memeID;
    @SerializedName("rid")
    private String reactorID;


    public static Reaction create(ReactionType type,String memeID){
        return new Reaction(type,memeID);
    }
    private Reaction(ReactionType type, String memeID) {
        this.type = type.ordinal();
        this.memeID = memeID;
    }
    public ReactionType getType() {
        return ReactionType.values()[type];
    }

    public String getMemeID() {
        return memeID;
    }

    public String getReactorID() {
        return reactorID;
    }
}

File: backend/src/main/java/com/memeit/backend/models/Reaction.kt
package com.memeit.backend.models

import android.os.Parcel
import android.os.Parcelable
import com.google.gson.annotations.SerializedName

/**
 * Created by Jv on 6/29/2018.
 */
data class Reaction(private var type: Int = 0,
                    @SerializedName("mid") var memeID: String? = null,
                    @SerializedName("rid") val reactorID: String? = null,
                    val poster: Poster? = null) : Parcelable {


    enum class ReactionType(val score: Float) {
        FUNNY(2f), VERY_FUNNY(3f), STUPID(-1f), ANGERING(-2f);

        fun create(mid: String?=null, rid: String? = null, poster: Poster? = null) =
                Reaction(ordinal, mid, rid, poster)
    }

    constructor(parcel: Parcel) : this(
            parcel.readInt(),
            parcel.readString(),
            parcel.readString(),
            parcel.readParcelable(Poster::class.java.classLoader))

    fun getType(): ReactionType {
        return ReactionType.values()[type]
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeInt(type)
        parcel.writeString(memeID)
        parcel.writeString(reactorID)
        parcel.writeParcelable(poster, flags)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<Reaction> {
        override fun createFromParcel(parcel: Parcel): Reaction {
            return Reaction(parcel)
        }

        override fun newArray(size: Int): Array<Reaction?> {
            return arrayOfNulls(size)
        }
    }
}

data class ReactionGroup(private val type: Int = 0, var count: Int = 0) {
    fun getType(): Reaction.ReactionType {
        return Reaction.ReactionType.values()[type]
    }
}
